sequencer/src/state.rs

```diff
@@ -20,11 +26,79 @@ impl HotShotState for State {
+                // This cannot fail, retry until we succeed.
+                loop {
+                    let retry_delay = Duration::from_millis(100);
+                    let chain_tip_number = match l1_provider.get_block_number().await {
+                        Ok(number) => number,
+                        Err(err) => {
+                            tracing::error!(
+                                "failed to get block number from L1 provider, retrying ({err})"
+                            );
+                            sleep(retry_delay).await;
+                            continue;
+                        }
+                    };
+                    tracing::debug!("L1 chain tip at block {chain_tip_number}");
+
+                    // Use a block with a few confirmations to minimize the probability of reorgs.
+                    let number = chain_tip_number.saturating_sub(3.into());
+
+                    let block = match l1_provider.get_block(number).await {
+                        Ok(Some(block)) => block,
+                        Ok(None) => {
+                            // This can only happen if there was a reorg that changed the block
+                            // height. Just retry.
+                            tracing::error!("L1 block {number} does not exist, retrying");
+                            sleep(retry_delay).await;
+                            continue;
+                        }
+                        Err(err) => {
+                            tracing::error!(
+                                "failed to get block {number} from L1 provider, retrying ({err})"
+                            );
+                            sleep(retry_delay).await;
+
+                            // Retry the whole loop, including re-fetching the block number. One
+                            // reason we might have failed here is if an L1 reorg changed the chain
+                            // height, and the block at our current `number` may not become
+                            // available again for a while.
+                            continue;
+                        }
+                    };
+
+                    let Some(number) = block.number else {
+                        // This can also happen if there's a reorg.
+                        tracing::error!("L1 block {number} is not committed, retrying");
+                        sleep(retry_delay).await;
+                        continue;
+                    };
+
+                    tracing::debug!(
+                        "proposing with L1 block {number} (timestamp {})",
+                        block.timestamp
+                    );
+                    break L1BlockInfo {
+                        number: number.as_u64(),
+                        timestamp: block.timestamp,
+                    };
+                }
+            })
+        } else {
+            // For unit testing, we may disable the L1 provider and use mock L1 blocks instead.
```

Comment:
Should we hide this behavior behind the `testing` feature flag?

Reply:
Yes, good idea